jsonpath-ast/src/ast.rs

```diff
@@ -0,0 +1,868 @@
+                        })?;
+                    }
+                    *pest = clone;
+                    Ok(this)
+                } else {
+                    Err(::from_pest::ConversionError::NoMatch)
+                }
+            }
+        }
+    };
+}
+
+use super::parse::{JSPathParser, Rule};
+#[cfg(feature = "compiled-path")]
+use crate::syn_parse::parse_impl::{
+    parse_bool, parse_float, validate_function_name, validate_js_int, validate_js_str,
+    validate_member_name_shorthand, ParseUtilsExt,
+};
+use derive_new::new;
+use from_pest::{ConversionError, FromPest, Void};
+use pest::iterators::{Pair, Pairs};
+use pest::Parser;
+use pest_ast::FromPest;
+use proc_macro2::Span;
+#[cfg(feature = "compiled-path")]
+use syn::parse::ParseStream;
+use syn::punctuated::Punctuated;
+use syn::token::Bracket;
+#[allow(unused_imports)]
+use syn::LitBool;
+use syn::{token, Ident, Token};
+#[cfg(feature = "compiled-path")]
+use syn_derive::Parse;
+
+pub trait KnowsRule {
+    const RULE: Rule;
+}
+
+#[derive(Debug, new, PartialEq)]
+pub struct PestIgnoredPunctuated<T, P>(pub(crate) Punctuated<T, P>);
+
+impl<'pest, T, P> FromPest<'pest> for PestIgnoredPunctuated<T, P>
+where
+    T: FromPest<'pest, Rule=Rule, FatalError=Void> + KnowsRule + std::fmt::Debug,
+    P: Default,
+{
+    type Rule = Rule;
+    type FatalError = Void;
+
+    fn from_pest(
+        pest: &mut Pairs<'pest, Self::Rule>,
+    ) -> Result<Self, ConversionError<Self::FatalError>> {
+        let parsed_items = Vec::<T>::from_pest(pest)?;
+
+        Ok(PestIgnoredPunctuated(Punctuated::from_iter(
+            parsed_items.into_iter(),
+        )))
+    }
+}
+
+/// Allows for syn to parse things that pest checks but does not store as rules
+#[derive(Debug, Default, new, PartialEq)]
+pub struct PestLiteralWithoutRule<T>(pub(crate) T);
+
+impl<T> From<T> for PestLiteralWithoutRule<T> {
+    fn from(value: T) -> Self {
+        Self(value)
+    }
+}
+impl<'pest, T: Default> FromPest<'pest> for PestLiteralWithoutRule<T> {
+    type Rule = Rule;
+    type FatalError = Void;
+
+    /// Always generates default value and leaves parse stream alone
+    fn from_pest(
+        _pest: &mut Pairs<'pest, Self::Rule>,
+    ) -> Result<Self, ConversionError<Self::FatalError>> {
+        Ok(PestLiteralWithoutRule(T::default()))
+    }
+}
+
+#[derive(Debug, new, PartialEq, FromPest)]
+#[cfg_attr(feature = "compiled-path", derive(Parse))]
+#[pest_ast(rule(Rule::main))]
+pub struct Main {
+    pub(crate) jp_query: JPQuery,
+    pub(crate) eoi: EOI,
+}
+
+impl Main {
+    /// Convenience function to allow tests to not import pest::parser::Parser
+    pub fn try_from_pest_parse(str: &str) -> Result<Self, ()> {
+        let mut rules = JSPathParser::parse(Rule::main, str).map_err(|_| ())?;
+        // *IF* the FromPest implementations are correctly written then Main::from_pest *cannot fail*
+        Main::from_pest(&mut rules).map_err(|_| ())
```

Comment:
i would leave the return error or introduce a new one. Now, we loose the info even when we don't need, it is stil better to have the error

---
ah, it is an intermdiaite state. The comment for the future final review then :)